aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,911 @@
+        // replica writes to the table at a time and that each replica gets a consistent
+        // view of the table state.
+        //
+        // ExclusiveLock does not conflict with the AccessShare lock taken by table
+        // reads, i.e. other aggregator replicas can continue to refresh their key
+        // caches without being blocked.
+        tx.lock_global_hpke_keypairs().await?;
+
+        let current_keypairs = tx.get_global_hpke_keypairs().await?;
+        debug!(?current_keypairs, "before");
+        let mut available_ids = Self::available_ids(&current_keypairs);
+        for config in &configs {
+            Self::run_hpke_for_config(tx, &current_keypairs, &mut available_ids, config).await?;
+        }
+
+        // Defensive assertion: Check our transaction snapshot for at least one active keypair in
+        // in the table. If one is absent, committing the transaction would leave Janus unstartable
+        // so we should rollback. This captures any bugs, and if the operator is trying to retire
+        // the last available key.
+        let current_keypairs = tx.get_global_hpke_keypairs().await?;
+        debug!(?current_keypairs, "after");
+        if !current_keypairs
+            .iter()
+            .any(|keypair| keypair.state() == &HpkeKeyState::Active)
+        {
+            Err(DatastoreError::User(
+                anyhow!("unexpected state: no keypairs are active").into(),
+            ))
+        } else {
+            Ok(())
+        }
+    }
+
+    #[tracing::instrument(err, skip(tx, current_keypairs, available_ids))]
+    async fn run_hpke_for_config(
+        tx: &Transaction<'_, C>,
+        current_keypairs: &[GlobalHpkeKeypair],
+        available_ids: &mut impl Iterator<Item = HpkeConfigId>,
+        config: &HpkeKeyRotatorConfig,
+    ) -> Result<(), DatastoreError> {
+        let clock = tx.clock();
+
+        let PartitionedKeypairs {
+            pending_keypairs,
+            active_keypairs,
+            expired_keypairs,
+        } = Self::partition_keypairs(current_keypairs, &config.ciphersuite);
+
+        let mut next_key = || {
+            generate_hpke_config_and_private_key(
+                available_ids
+                    .next()
+                    .ok_or_else(|| DatastoreError::User(anyhow!("u8 space exhausted").into()))?,
+                config.ciphersuite.kem_id(),
+                config.ciphersuite.kdf_id(),
+                config.ciphersuite.aead_id(),
+            )
+            .map_err(|e| DatastoreError::User(e.into()))
+        };
+
+        if config.retire {
+            for pending_keypair in pending_keypairs {
+                // Janus replicas should have never advertised this keypair, so it should be safe
+                // to delete outright.
+                info!(id = ?pending_keypair.id(), "deleting pending keypair for retired ciphersuite");
+                tx.delete_global_hpke_keypair(pending_keypair.id()).await?;
+            }
```

Comment:
n.b. we could `try_join` a lot of calls like this, but since this code is not performance sensitive, I prioritized readability and straightforwardness.